Fixes

FriskIsGit opened the pull request.

FriskIsGit committed:
Initial commit

FriskIsGit committed:
Changed full paths to shorter xpaths. Fixed displayPositionRelativeToWindow, movePointerWithinWindowBy. Added initiateDisplayThread to facilitate automation.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        new AutomateBrowserActions().customizeBrowser();
+    }
+}
+class Cursor{
+    private int cursorXPosition = 0;
+    private int cursorYPosition = 0;
+
+    public void setCursorXPosition(int cursorXPosition) {
+        this.cursorXPosition = cursorXPosition;
+    }
+    public void setCursorYPosition(int cursorYPosition) {
+        this.cursorYPosition = cursorYPosition;
+    }
+
+    public int getCursorXPosition() {
+        return cursorXPosition;
+    }
+    public int getCursorYPosition() {
+        return cursorYPosition;
+    }
+}
```
This should be a separate class, probably in a separate file.

FriskIsGit replied:
1 file 1 class rule?

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        new AutomateBrowserActions().customizeBrowser();
+    }
+}
+class Cursor{
```
Such a class (only private fields with getters and setters) is called a DTO - Data Transfer Object. Do not implement this by hand, use Lombok instead - specifically `@Data` annotation.

Doing so will help you refactor this class into:
```
@Data class CursorPosition {
  private int x;
  private int y;
}
```

FriskIsGit replied:
IntelliJ refactoring feature is more intelligent than I thought
I feel like Lombok was made for lazy coders :P

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        new AutomateBrowserActions().customizeBrowser();
+    }
+}
+class Cursor{
```
In general this class is not a `Cursor` implementation - it is a cursor position at best. I'd rename class and it's fields to `CursorPosition`, `x` and `y` respectively.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        return cursorYPosition;
+    }
+}
+class Performer {
```
This class is an unnecessary complication. Consider deleting it and using `Actions` directly. 

Line 18 would become:
```
private Actions actions;
```
Line 24:
```
actions = new Actions(driver);
```
And usage (e.g. line 43) will be:
```
actions.sendKeys(Keys.ARROW_DOWN).perform();
```

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+import java.util.List;
+
+class AutomateBrowserActions {
+    java.awt.Dimension screenDimensions = Toolkit.getDefaultToolkit().getScreenSize();
```
This field is used only in a single method. Consider defining a variable in that method instead of defining a field.

FriskIsGit replied:
I changed it so that it's called whenever a function needs to resize the window to execute properly

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+    private final long timeInMs = 60000;
+
+    private void customizeBrowser(){
+        System.setProperty("webdriver.chrome.driver", "src/main/resources/chromedriver.exe");
```
Consider defining properties in [POM properties](https://maven.apache.org/pom.html#properties).

FriskIsGit replied:
this proved to be more difficult than I thought

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+    private void customizeBrowser(){
+        System.setProperty("webdriver.chrome.driver", "src/main/resources/chromedriver.exe");
+        WebDriver driver = new ChromeDriver();
+        performer = new Performer(driver);
```
Such initialization code should be placed in class constructor. Consider adding:
```
...
private AutomateBrowserActions () {
  cursor = new Cursor()
  driver = new ChromeDriver();
  performer = new Performer(driver);
}
...
```
and removing these two lines from this method.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+    private Performer performer;
+    private final long timeInMs = 60000;
+
+    private void customizeBrowser(){
```
This method does not increase readability - in fact it decreases it slightly.

Compare current version:
```
    public static void main(String[] args) {
        new AutomateBrowserActions().customizeBrowser();
    }
```
with this method inlined
```
    public static void main(String[] args) {
        BrowserActions browserActions = new AutomateBrowserActions();

        browserActions.resizeAndRepositionWindow();
        browserActions.initiateDisplayThread();
        browserActions.denyGoogleCookies();
    }
```
The latter is easier to understand and maintain.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        //denyCameraAccess(driver);
+    }
+
+    private void changeBrowserLanguage(WebDriver driver){
```
Most of the methods in this class take `WebDriver` as an argument - it should be defined as a field and set up in the constructor.

FriskIsGit replied:
should there be a getter for WebDriver field?

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+    private final int windowWidth = 960;
+    private final int windowHeight = 720;
+    private final Cursor cursor = new Cursor();
+    private Performer performer;
```
This can probably be final as well.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+    private final int windowHeight = 720;
+    private final Cursor cursor = new Cursor();
+    private Performer performer;
+    private final long timeInMs = 60000;
```
Consider grouping static values and non-static fields, it increases readability:
```
private static final int WINDOW_WIDTH = 960;
private static final int WINDOW_HEIGHT = 720;
private static final int TIME_IN_MS = 60000;

private final Cursor cursor;
private final Performer performer;
```

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+class AutomateBrowserActions {
+    java.awt.Dimension screenDimensions = Toolkit.getDefaultToolkit().getScreenSize();
+    //4:3 window - minimal dimensions
+    private final int windowWidth = 960;
```
Should be `static` - there is no reason to maintain that field on per-object basis. Same for other constant integers.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        initiateDisplayThread(driver);
+        denyGoogleCookies(driver);
+        //denyYahooCookies(driver);
+        //denyCameraAccess(driver);
```
Code that's commented out can be safely removed - less maintenance over time.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        //denyCameraAccess(driver);
+    }
+
+    private void changeBrowserLanguage(WebDriver driver){
```
This method can be replaced by starting browser with the correct language in the first place - https://stackoverflow.com/a/18646033/1786019.

Current implementation is very prone to issues like Google re-designing their web interface and changing the position of your preferred language in the UI.

FriskIsGit replied:
This worked, surprisingly well.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        clickPointerWithinWindow();
+        clickPointerWithinWindow();
+    }
+    private void changeDefaultSearchEngineTo(WebDriver driver,int element) {
```
Out of curiosity, what's the purpose of that method? In selenium we navigate to pages using their explicit URLs, so we can navigate to bing or duckduckgo if we want to search through those. Implementation is reaaaally hard to maintain in the current form - I expect these randon numbers like 36, 0.25 and 25 to work only on some specific screen/browser/system.

FriskIsGit replied:
For the sake of somehow bypassing google settings hidden in #shadow-root
These numbers are compatible with 3.141.59 version of chrome, current google UI and provided window dimensions i.e. screen resolution must be either equal or larger than given 960 by 720

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        performer.getActionsBuilder().sendKeys(Keys.RETURN).perform();
+    }
+
+    private void denyLocationAccess(WebDriver driver){
```
AFAIK this can be replaced by a custom property - https://stackoverflow.com/a/21077039/1786019.

FriskIsGit replied:
Doesn't change the defaults

FriskIsGit replied:
JSONObject solution or any map solution have no effect on location settings

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        movePointerWithinWindowBy((int) (windowWidth*0.5),155);
+        clickPointerWithinWindow();
+    }
+    private void denyCameraAccess(WebDriver driver){
```
Can be done through prefs as well I believe - https://stackoverflow.com/a/47836305/1786019.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        wait.until(ExpectedConditions.visibilityOf(customizePrivacySettingsElement));
+        customizePrivacySettingsElement.click();
+
+        List<WebElement> listOfWebElements = driver.findElements(By.xpath("//div[@class='uScs5d']//div[@data-is-touch-wrapper='true']"));
```
Can we get a better name for the list of web elements? What are these? Cookies check boxes? Something else?

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+    private void denyGoogleCookies(WebDriver driver){
+        WebDriverWait wait= new WebDriverWait(driver,15);
+        driver.get("https://www.google.com");
+        // //div[@class='uScs5d']//div[@data-is-touch-wrapper='true']
```
Remove commented out code.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+    private void clickPointerWithinWindow(){
+        performer.getActionsBuilder().click().perform();
+    }
+    private void initiateDisplayThread(WebDriver driver) {
```
Consider some other method name, better reflecting the purpose of this - `logCursorPositionOverTime` or `continuouslyLogCursorPosition` maybe?

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        });
+        displayThread.start();
+    }
+    private void clickOnElement(WebElement webElement){
```
Such a method does not increase readability, consider inlining.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        try {
+            Thread.sleep(time);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```
This isn't a good programming practice to ignore runtime exceptions. Consider adding `throws InterruptedException` to methods which need to use sleep and use `Thread.sleep(time)` directly.

Or even better - consider not sleeping at all.

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        driver.manage().window().setSize(new Dimension(windowWidth, windowHeight));
+        driver.manage().window().setPosition(new Point(screenDimensions.width-windowWidth, 0));
+    }
+    private void movePointerWithinWindowBy(int x,int y){
```
We're calling it cursor everywhere else so maybe `moveCursorWithingWindowBy`?

karolgil commented on src/main/java/AutomateBrowserActions.java:
```diff
@@ -0,0 +1,200 @@
+        }
+        performer.getActionsBuilder().moveByOffset(x,y).build().perform();
+    }
+    private void clickPointerWithinWindow(){
```
Maybe just `click`?

FriskIsGit committed:
restructured code
removed unnecessary funcs&variables
added denyYahooCookies func

await new solutions

FriskIsGit committed:
replaced changeBrowserLanguage func with proper settings argument which caused chrome UI to change shape which as a result made me delete other methods cuz they didnt work anymore ;-|
